File: cere/cere_test/clusterize_invocations.R
#!/usr/bin/env Rscript

args <- commandArgs(trailingOnly = TRUE)

suppressPackageStartupMessages(require(ggplot2, quietly=TRUE))
suppressPackageStartupMessages(require(reshape2, quietly=TRUE))
suppressPackageStartupMessages(require(data.table, quietly=TRUE))
suppressPackageStartupMessages(require(plyr, quietly=TRUE))
suppressPackageStartupMessages(require(GMD, quietly=TRUE))
suppressPackageStartupMessages(require(fpc, quietly=TRUE))
suppressPackageStartupMessages(require(cluster, quietly=TRUE))
suppressPackageStartupMessages(require(fields, quietly=TRUE))
options(warn=-1)
options(error=traceback)
MAX_POINTS=50000
ROOT_PLOTS = "cere_measures/plots/"
PALETTE = c("red","blue","green","yellow","black","grey","pink","maroon","orange","purple","magenta",
            "silver","golden","brown","cyan")
set.seed(2000)

# Check the arguments
if ((length(args) < 3)) {
    print("usage: ./plot_codelets.R <codelet_name> <csv file> <bynary files>")
    q()
}

plot_by_Codelets <- function(tracedLoop, CodeletName) {
    p <- ggplot(tracedLoop, aes(y=cycles, x=invocation))
    p <- p + geom_point()
    p <- p + ggtitle(CodeletName)
    ggsave(file="test.png", plot=p, dpi=300)
}

plot_by_Cluster <- function(table, CodeletName) {
    p <- ggplot(table, aes(y=cycles, x=invocation, colour=as.character(Cluster), shape = is.representative))
    p <- p + geom_point()
    p <- p + theme_bw() + ylab("Cycles")
    p <- p + scale_colour_manual(values=PALETTE)
    p <- p + theme(legend.position="none")
    ggsave(file= paste(ROOT_PLOTS,CodeletName,"_byPhase.png",sep=""), plot=p, height=4.16, width=10, dpi=300)
}

# load_csv: reads a csv file
load_csv <- function(csvFile) {
    return (tryCatch(
                read.table(csvFile, header=T, comment.char = "", sep=","),
                error = function(e) {
                        print(paste("Could not open file", csvFile))
                        q()
                    }
                ))
}

dump_to_file <- function(todump, filename) {
    write.table(todump, filename, sep=" ", quote = F, row.names = F, col.names = F)
}

plot_withinss <- function(allValues) {
    wss <- numeric(15)
    for (i in 1:15) wss[i] <- sum(kmeans(allValues$cycles,
                                       centers=i)$withinss)
    return(wss)
}

clust.centroid = function(i, dat, clusters) {
  ind = (clusters == i)
  return(median(dat[ind, drop=FALSE]))
}

clust.sum = function(i, dat, clusters) {
  ind = (clusters == i)
  return(sum(as.numeric(dat[ind, drop=FALSE])))
}

clust.length = function(i, dat, clusters) {
  ind = (clusters == i)
  return(length(dat[ind, drop=FALSE]))
}

cluster <- function (allValues) {
    #if the codelet has less than 11 invocations
    if(nrow(allValues) <= 10)
    {
        nb_clust=nrow(allValues)
        print(paste("Nb Clusters = ", nb_clust, sep=""))
        return(seq(1:nb_clust))
    } else {
        pamk.best <- pamk(allValues$cycles, usepam=FALSE, scaling=TRUE, diss=FALSE)
        nb_clust=pamk.best$nc
        print(paste("Nb Clusters = ", nb_clust, sep=""))
        clusters = clara(allValues$cycles, nb_clust, metric = "ch", samples=50)
        return(clusters$clustering)
    }
}

CodeletName = args[1]
print(CodeletName)
nonTraceCycles = load_csv(args[2])

inVivoCycles=nonTraceCycles[nonTraceCycles$Codelet.Name==CodeletName, ]$CPU_CLK_UNHALTED_CORE
total_invocation=nonTraceCycles[nonTraceCycles$Codelet.Name==CodeletName, ]$Call.Count

binFile=args[3]
to.read=file(binFile, "rb")
tracedLoop=readBin(to.read, double(), n=total_invocation*2)
cycles <- tracedLoop[seq(1,total_invocation*2, 2)]

invocation=seq(1,total_invocation)
allValues = data.frame(cycles, invocation)

print(paste("Cycles = ", inVivoCycles))
print(paste("Invocations=", total_invocation))

POINTS_TO_KEEP=min(MAX_POINTS, total_invocation)

if (total_invocation > MAX_POINTS) {
    allValues=allValues[sample(nrow(allValues), POINTS_TO_KEEP, replace=F), ]
}

total_kept_cycles = sum(allValues$cycles)

#Clusterize data
clusters = cluster(allValues)
allValues = cbind(allValues, Cluster=clusters)
#Compute centroids of each cluster
Centroids <- sapply(sort(unique(allValues$Cluster)), clust.centroid, allValues$cycles, allValues$Cluster)

ClusterInfo = data.frame(Cluster=1:length(Centroids), Centroid = Centroids)

# Merge clusters
find_merge <- function() {
    if (nrow(ClusterInfo) == 1) { return(list())}
    for (i in 1:nrow(ClusterInfo)) {
        for (j in i:nrow(ClusterInfo)) {
            if (j != i) {
                left = ClusterInfo[i,]$Centroid
                right = ClusterInfo[j,]$Centroid
                error = abs(left-right)/max(left,right)
                if (error <= 0.01) {
                    # merge left and right clusters
                    return(list(ClusterInfo[i,]$Cluster, ClusterInfo[j,]$Cluster))
                }
            }
        }
    }
    return(list())
}

while(1) {
    merged = find_merge()
    if (length(merged) == 0) break;
    left = merged[[1]]
    right = merged[[2]]
    ClusterInfo = ClusterInfo[ClusterInfo$Cluster != right,]
    allValues[allValues$Cluster==right,]$Cluster = left
}

#Compute for each invocation, the distance to its cluster centroid.
#And elect as representative invocation, the closest to the centroid.
allValues$is.representative=F
for (clust in unique(allValues$Cluster)) {
    tmp = allValues[allValues$Cluster==clust, ]
    tmp$Cluster <- NULL
    center = ClusterInfo[ClusterInfo$Cluster == clust, ]$Centroid

    tmp$DistToCentroid = abs(tmp$cycles - center)
    tmp <- tmp[order(tmp$DistToCentroid), ] #order first by distance to centroid

    #We select as representative the closest invocation to the centroid
    allValues[allValues$Cluster==clust, ]$is.representative =
      ifelse(allValues[allValues$Cluster==clust, ]$invocation == tmp[1, ]$invocation, T, F)
}

#compute distance between clusters
distances = rdist(ClusterInfo$Centroid)

#Find the nearest cluster to clust
find_nearest_cluster <- function(clust, distances) {
    tmp_vec = distances[,clust]
    tmp_vec[[clust]] = Inf
    tmp_vec <- order(tmp_vec)
    nearest_clust=tmp_vec[[1]]
    i=1
    while(nearest_clust != clust) {
        #We have to check if that cluster has a representative > 2000 cycles
        if (allValues[allValues$Cluster==clust & allValues$is.representative==T, ]$cycles >= 2000) {
            return(nearest_clust)
        }
        i=i+1
        nearest_clust = tmp_vec[[i]]
    }
    return(0)
}

#If a representative is below 2000 cycles, we search
#for the nearest cluster to fused them both.
if (nrow(ClusterInfo) != 1) {
    for (i in 1:nrow(ClusterInfo)) {
        clust=ClusterInfo[i,]$Cluster
        if (allValues[allValues$Cluster==clust & allValues$is.representative==T, ]$cycles < 2000) {
            nearest_clust = find_nearest_cluster(clust, distances)
            if(nearest_clust != 0) {
                allValues[allValues$Cluster==clust, ]$is.representative = F
                allValues[allValues$Cluster==clust, ]$Cluster = nearest_clust
            }
        }
    }
}

#Compute for each cluster its size
Sums <- sapply(sort(unique(allValues$Cluster)), clust.sum, allValues$cycles, allValues$Cluster)

plot_by_Cluster(allValues, CodeletName)

res <- allValues[allValues$is.representative==T, ]
res <- res[order(res$Cluster), ]

#The part of each cluster time in the total loop time.
res$part = (Sums/res$cycles) * inVivoCycles/total_kept_cycles
print(res)
todump=data.frame(res$invocation, res$part, res$cycles)
dump_to_file(todump, paste(CodeletName, ".invocations", sep=""))

File: cere/cere_test/clusterize_invocations.py
#!/usr/bin/env python
from __future__ import print_function
import csv
import numpy as np
import matplotlib
matplotlib.use('Agg')

import matplotlib.pyplot as plt
from sklearn import cluster
from sklearn.preprocessing import StandardScaler


MAX_POINTS=50000

PALETTE = ["red","blue","green","yellow","black","grey","pink",
            "maroon","orange","purple","magenta","silver","golden",
            "brown","cyan"]

def parse_codelet_csv(csvfile, codelet):
    with open(csvfile) as f:
        reader = csv.DictReader(f)
        try:
            entry = reader.next()
            assert(entry['Codelet Name'] == codelet)
            return entry
        except StopIteration:
            print("could not parse {csv} file".format(csv),
                  file=sys.stderr)
            sys.exit(1)

def parse_trace_file(tracefile):
    trace = np.fromfile(tracefile, dtype='d', count=-1)

    # filter the trace:
    # 1) odd items are cycles
    # 2) even items are invocation numbers
    size = len(trace) / 2
    return dict (cycles = trace[::2], invocations = trace[1::2],
                 size = size)

def subsample(trace, n):
    samples = np.random.choice(trace['size'], n)
    trace['invocations'] = trace['invocations'][samples,]
    trace['cycles'] = trace['cycles'][samples,]

def clusterize(trace):
    min_samples = max(1,trace['size']/1000)
    cycles = np.reshape(trace['cycles'], (trace['size'],1))
    cycles = StandardScaler().fit_transform(cycles)
    clusterer = cluster.DBSCAN(min_samples=min_samples, eps=.3)
    clusterer.fit(cycles)
    return clusterer.labels_

def clusterize_invocations(codelet, csvfile, tracefile):
    codelet_info = parse_codelet_csv(csvfile, codelet)
    in_vivo_cycles = float(codelet_info['CPU_CLK_UNHALTED_CORE'])

    trace = parse_trace_file(tracefile)

    # subsample trace if needed
    if trace['size'] > MAX_POINTS:
        subsample(trace, MAX_POINTS)

    # label the data using DBSCAN clusterer
    labels = clusterize(trace)


    # take all clusters except 'noise' cluster
    clusters = [ c for c in np.sort(np.unique(labels)) if c >= 0 ]

    # compute total weight for non-noise points
    total_weight = np.sum(trace['cycles'][labels >= 0])

    representatives = []
    weights = []
    repcycles = []

    # find representatives
    for c in clusters:
        inside_cluster = (labels == c)
        points = trace['cycles'][inside_cluster]
        invocation = trace['invocations'][inside_cluster]
        center = np.median(points)

        # representative is the closest invocation to the center
        repind = np.argmin(abs(points - center))
        rep = int(invocation[repind])
        representatives.append(rep)

        # cycles of the representative
        repcycles.append(points[repind])

        # weight of the cluster
        wei = (np.sum(points) / points[repind] *
               in_vivo_cycles  / total_weight)
        weights.append(wei)


    with file(codelet + '.invocations', 'w') as output:
        for i in range(len(clusters)):
            line = "{invocation} {weight} {cycles}\n".format(
                invocation = representatives[i],
                weight = weights[i],
                cycles = repcycles[i])
            output.write(line)

    #plot clusters
    f, ax = plt.subplots(1, figsize=(10, 4.16))
    for c, color in zip(clusters, PALETTE):
        inside_cluster = (labels == c)
        points = trace['cycles'][inside_cluster]
        invocations = trace['invocations'][inside_cluster]

        ax.plot(invocations, points, 'o', markerfacecolor=color,
                 markeredgecolor=color, markersize=4)

    ax.plot(trace['invocations'][labels < 0],
             trace['cycles'][labels < 0],
             'x', markerfacecolor='lightgray',
             markeredgecolor='lightgray', markersize=4)


    ax.grid('on')
    ax.set_ylabel('cycles')
    ax.set_xlabel('invocation')
    f.savefig('cere_measures/plots/' + codelet + '_byPhase.png',
            bbox_inches='tight', dpi=100)


if __name__ == "__main__":
    import sys
    if len(sys.argv) != 4:
        print("usage: clusterize_invocations.py <codelet_name> <csv> <bin>",
              file=sys.stderr)
        sys.exit(1)

    clusterize_invocations(sys.argv[1], sys.argv[2], sys.argv[3])
